src/header/common/range.rs

```diff
@@ -0,0 +1,231 @@
+/// other-ranges-specifier = other-range-unit "=" other-range-set
+/// other-range-set = 1*VCHAR
+/// ```
+///
+/// # Example values
+/// * `bytes=1000-`
+/// * `bytes=-2000`
+/// * `bytes=0-1,30-40`
+/// * `custom_unit=0-123,-200`
+///
+/// # Examples
+/// ```
+/// use hyper::header::{Headers, Range, RangeSpec, RangeUnit};
+///
+/// let mut headers = Headers::new();
+///
+/// headers.set(Range {
+///     unit: RangeUnit::Bytes,
+///     ranges: vec![RangeSpec::FromTo(1, 100), RangeSpec::AllFrom(200)]
+/// });
+/// ```
+/// ```
+/// use hyper::header::{Headers, Range};
+///
+/// let mut headers = Headers::new();
+/// headers.set(Range::bytes(1, 100));
+/// ```
+#[derive(PartialEq, Clone, Debug)]
+pub struct Range {
+    /// Unit of the Range i.e. bytes
+    pub unit: RangeUnit,
+    /// Set of ranges as defined in the HTTP spec
+    pub ranges: Vec<RangeSpec>,
+}
+
+/// Each 'Range' header can contain one or more RangeSpecs.
+/// Each RangeSpec defines a range of units to fetch
+#[derive(PartialEq, Clone, Debug)]
+pub enum RangeSpec {
+    /// Get all bytes between x and y ("x-y")
+    FromTo(u64, u64),
+    /// Get all bytes starting from x ("x-")
+    AllFrom(u64),
+    /// Get last x bytes ("-x")
+    Last(u64)
+}
+
+impl Range {
+    /// Get the most common byte range header ("bytes=from-to")
+    pub fn bytes(from: u64, to: u64) -> Range {
+        Range {
+            unit: RangeUnit::Bytes,
+            ranges: vec![RangeSpec::FromTo(from, to)],
+        }
+    }
+}
+
+
+impl fmt::Display for RangeSpec {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            RangeSpec::FromTo(from, to) => write!(fmt, "{}-{}", from, to),
+            RangeSpec::Last(pos) => write!(fmt, "-{}", pos),
+            RangeSpec::AllFrom(pos) => write!(fmt, "{}-", pos),
+        }
+    }
+}
+
+
+impl fmt::Display for Range {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(f, "{}=", self.unit));
+
+        for (i, range) in self.ranges.iter().enumerate() {
+            if i != 0 {
+                try!(f.write_str(","));
+            }
+            try!(Display::fmt(range, f));
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for Range {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Range, ()> {
+        let mut iter = s.splitn(2, "=");
+
+        match (iter.next(), iter.next()) {
+            (Some(unit), Some(ranges)) => {
+                match (RangeUnit::from_str(unit), from_one_comma_delimited(ranges.as_bytes())) {
+                    (Ok(unit), Ok(ranges)) => {
+                        if ranges.is_empty() {
+                            return Err(());
+                        }
+                        Ok(Range{unit: unit, ranges: ranges})
+                    },
+                    _ => Err(())
+                }
+            }
+            _ => Err(())
+        }
+    }
+}
+
+impl FromStr for RangeSpec {
+    type Err = ();
```

Comment:
I'm not sure if `hyper::Error` makes sense here. At least not `hyper::Error::Header`. The `RangeSpec` struct isn't really a header, so `::Error::Header` doesn't make much sense for parsing failure. Plus I don't see any other `::Error` enum value, that would make sense in this context.